### PR TITLE
add github:environment:create and github:deployKey:create

### PR DESCRIPTION
**Pull request by aochsner**
## Hey, I just made a Pull Request!

Closes: #17515 

#### :heavy_check_mark: Checklist

- [x] A changeset describing the change and affected packages. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#creating-changesets))
- [x] Added or updated documentation
- [x] Tests for new functionality and regression tests for bug fixes
- [x] Screenshots attached (for UI changes)
- [x] All your commits have a `Signed-off-by` line in the message. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#developer-certificate-of-origin))

### CODE REVIEW

**Comment by aochsner** [comment]
Not sure I understand how to fix the error in the workflow...

**Comment by aochsner** [comment]
It seems the failing E2E tests have been failing for a while... don't believe I caused this... https://github.com/backstage/backstage/actions/workflows/verify_e2e-linux.yml

not sure how to fix it...

**Comment by benjdlambert** [comment]
The issues on the main branch have been resolved, you should be able to rebase to get the fixes.

**Comment by aochsner** [comment]
OK.  All good to go!

File: plugins/github-issues/package.json [hunk 1]
```diff
@@ -33,7 +33,7 @@
     "@material-ui/icons": "^4.9.1",
     "@material-ui/lab": "4.0.0-alpha.61",
     "luxon": "^3.0.0",
-    "octokit": "^2.0.4",
+    "octokit": "^2.0.14",
```
Comment (freben) [inline plugins/github-issues/package.json:36]: So maybe possible to revert.

File: .changeset/plenty-suns-check.md [hunk 2]
```diff
@@ -0,0 +1,5 @@
+---
+'@backstage/plugin-github-issues': patch
+---
+
+Bump octokit version
```
Comment (freben) [inline .changeset/plenty-suns-check.md:5]: Same as in #17722 right?

File: plugins/scaffolder-backend/src/scaffolder/actions/builtin/github/helpers.ts [hunk 3]
```diff
@@ -146,7 +146,8 @@ export async function createGithubRepoWithCollaboratorsAndTopics(
           name: repo,
           org: owner,
           private: repoVisibility === 'private',
-          visibility: repoVisibility,
+          visibility:
+            repoVisibility === 'internal' ? undefined : repoVisibility,
```
Comment (freben) [inline plugins/scaffolder-backend/src/scaffolder/actions/builtin/github/helpers.ts:150]: This looks unrelated - was the `internal` value actually not valid for the API?

File: plugins/scaffolder-backend/src/scaffolder/actions/builtin/github/githubEnvironment.ts [hunk 4]
```diff
@@ -0,0 +1,206 @@
...
+            owner: owner,
+            repo: repo,
+            environment_name: name,
+            name: item,
+          });
+        }
+      }
+
+      for (const key in environmentVariables) {
+        if ({}.hasOwnProperty.call(environmentVariables, key)) {
```
Comment (freben) [inline plugins/scaffolder-backend/src/scaffolder/actions/builtin/github/githubEnvironment.ts:161]: Same as in #17722 - would prefer `for (const [key, value] of Object.entries(...))` if possible, same below

File: plugins/scaffolder-backend/package.json [hunk 5]
```diff
@@ -81,12 +81,13 @@
...
     "jsonschema": "^1.2.6",
     "knex": "^2.0.0",
+    "libsodium-wrappers": "^0.7.11",
     "lodash": "^4.17.21",
     "luxon": "^3.0.0",
     "morgan": "^1.10.0",
     "node-fetch": "^2.6.7",
     "nunjucks": "^3.2.3",
-    "octokit": "^2.0.0",
+    "octokit": "^2.0.14",
```
Comment (freben) [inline plugins/scaffolder-backend/package.json:90]: revert?

File: plugins/scaffolder-backend/src/scaffolder/actions/builtin/github/githubDeployKey.ts [hunk 6]
```diff
@@ -0,0 +1,159 @@
...
+import { ScmIntegrationRegistry } from '@backstage/integration';
+import { parseRepoUrl } from '../publish/util';
+import { getOctokitOptions } from './helpers';
+import { Octokit } from 'octokit';
+import Sodium from 'libsodium-wrappers';
+
+/**
+ * Creates an `github:api:execute` Scaffolder action.
+ *
+ * @remarks
```
Comment (freben) [inline plugins/scaffolder-backend/src/scaffolder/actions/builtin/github/githubDeployKey.ts:28]: remove

File: plugins/scaffolder-backend/src/scaffolder/actions/builtin/github/githubDeployKey.ts [hunk 7]
```diff
@@ -0,0 +1,159 @@
...
+import { InputError } from '@backstage/errors';
+import { createTemplateAction } from '@backstage/plugin-scaffolder-node';
+import { ScmIntegrationRegistry } from '@backstage/integration';
+import { parseRepoUrl } from '../publish/util';
+import { getOctokitOptions } from './helpers';
+import { Octokit } from 'octokit';
+import Sodium from 'libsodium-wrappers';
+
+/**
+ * Creates an `github:api:execute` Scaffolder action.
```
Comment (freben) [inline plugins/scaffolder-backend/src/scaffolder/actions/builtin/github/githubDeployKey.ts:26]: Not really :)

File: plugins/scaffolder-backend/src/scaffolder/actions/builtin/github/githubEnvironment.ts [hunk 8]
```diff
@@ -0,0 +1,206 @@
...
+import { InputError } from '@backstage/errors';
+import { createTemplateAction } from '@backstage/plugin-scaffolder-node';
+import { ScmIntegrationRegistry } from '@backstage/integration';
+import { parseRepoUrl } from '../publish/util';
+import { getOctokitOptions } from './helpers';
+import { Octokit } from 'octokit';
+import Sodium from 'libsodium-wrappers';
+
+/**
+ * Creates an `github:api:execute` Scaffolder action.
```
Comment (freben) [inline plugins/scaffolder-backend/src/scaffolder/actions/builtin/github/githubEnvironment.ts:26]: Not really :)

File: plugins/scaffolder-backend/src/scaffolder/actions/builtin/github/githubEnvironment.ts [hunk 9]
```diff
@@ -0,0 +1,206 @@
...
+ *
+ * @remarks
```
Comment (freben) [inline plugins/scaffolder-backend/src/scaffolder/actions/builtin/github/githubEnvironment.ts:28]: Remove

File: plugins/scaffolder-backend/src/scaffolder/actions/builtin/github/githubDeployKey.ts [hunk 10]
```diff
@@ -0,0 +1,159 @@
...
+    async handler(ctx) {
+      const {
+        repoUrl,
+        publicKey,
+        privateKey,
+        deployKeyName,
+        privateKeySecretName = `${deployKeyName
+          .split(' ')
+          .join('_')
+          .toLocaleUpperCase()}_PRIVATE_KEY`,
```
Comment (freben) [inline plugins/scaffolder-backend/src/scaffolder/actions/builtin/github/githubDeployKey.ts:104]:
```suggestion
          .toLocaleUpperCase('en-US')}_PRIVATE_KEY`,
```

you will get odd effects in some locales otherwise ...

**Review by freben — COMMENTED** [review]
Alright!

**Review by freben — COMMENTED** [review]
Alright!

File: plugins/scaffolder-backend/package.json (hunk 5 above)
Comment (aochsner) [reply plugins/scaffolder-backend/package.json]: i could have sworn this one was required but let me revert and give it another go.  Too many plates spinning at the same time (and some unlucky timing)

File: plugins/scaffolder-backend/src/scaffolder/actions/builtin/github/helpers.ts (hunk 3 above)
Comment (aochsner) [reply plugins/scaffolder-backend/src/scaffolder/actions/builtin/github/helpers.ts]: i swear typescript was telling me the api changed but again... too many plates spinning so let me revert and see what's really what.

**Comment by freben** [comment]
Noting that this is all red still.

**Comment by aochsner** [comment]
> Noting that this is all red still.

Thanks.  Just got back from holiday.  And was waiting for #17722 to merge (thanks!). Will start updating this accordingly

File: plugins/scaffolder-backend/api-report.md [hunk 11]
```diff
@@ -211,6 +211,50 @@ export function createGithubActionsDispatchAction(options: {
...
+export function createGithubEnvironmentAction(options: {
+  integrations: ScmIntegrationRegistry;
+}): TemplateAction_2<
+  {
+    repoUrl: string;
+    name: string;
+    deploymentBranchPolicy?:
+      | {
+          protected_branches: boolean;
+          custom_branch_policies: boolean;
```
Comment (freben) [inline plugins/scaffolder-backend/api-report.md:239]: Just one final question on this setup. These bools are required. Do they have to be? I'm just thinking that maybe it's hard to add more fields to this without breaking everybody's templates, if they keep being required.

**Review by freben — APPROVED** [review]
Let's go!

File: plugins/scaffolder-backend/api-report.md (hunk 11 above)
Comment (aochsner) [reply plugins/scaffolder-backend/api-report.md:239]: The field `deploymentBranchPolicy` is optional but if you specify it, you have to specify both variables, and they're mutually exclusive.  Kind of a kludgy API IMHO but just reflecting how Gitthub's API exposes it.

Maybe I'm missing your point tho...

Comment (freben) [reply plugins/scaffolder-backend/api-report.md:239]: No if it mirrors the actual underlying API then it is what it is, thanks for reporting back!
